pubtools/pulplib/_impl/fake/match.py

```diff
@@ -178,7 +203,7 @@ def convert_field_to_pulp(obj, field_name, value, subfield):
     if cls_field.type is datetime.datetime and isinstance(value, datetime.datetime):
         # We can convert from datetime back to the ISO8601 timestamp format
         # used within Pulp
-        return value.strftime("%Y-%m-%dT%H:%M:%SZ")
+        return value
```

Comment:
Either we should get rid of the if condition here, as 'value' is anyways being returned in the end OR avoid converting the matcher.\_value in the match\_* methods above.
I'll prefer the later and keeping this as-is to keep the matcher.\_value consistent with the definition rather than modifying it somewhere midway.
I suggest the transformed datetime object in the match\_* above is used there as required. That's the only place the transformed value is required as of now. Updating it to matcher.\_value may only add confusion as same object when passed through different matcher has a different value e.g. match\_field\_less changes the value but match\_in doesn't.
Also, the pulp\_dict in pulp\_value() looks more in line with how the objects exist in Pulp and this will keep it consistent.

Reply:
The problem ran into was convert_field_to_pulp was converting a datetime to a string when if you access the field directly it returns it as datetime. Creating some weirdness between "notes.created" and "created_date", "notes.created" was returning a string where "created_date" was returning a datetime.

We tested this against the real Matcher in pulp to ensure that this actually represents what is going on in pulp. I agree removing the if statement entirely though (because it doesn't make sense to have it there).

Reply:
I see what you mean. That's how Pulp works as well, converting the str to datetime and compare if provided the $date operator, which we usually do with datetime type values.
So, It's the first one then, getting rid of the condition :)
I'll suggest to add this transformation to match\_in matcher as well just to keep it consistent, though we don't have a use case yet matching multiple datetime values.

Reply:
All fixed up :)
Yeah I don't want this code to be this way.. but it matches what was seen in testing :(